File: src/core/portal.py
"""Parsing helpers for the GUET Dr.COM portal page.

The portal exposes session details and logout configuration as JavaScript
variables. This module converts those values into typed application state
without relying on account credentials or reading a password from the page.
"""

from __future__ import annotations

import enum
import re
from dataclasses import dataclass
from urllib.parse import urlsplit, urlunsplit


PORTAL_OPERATOR_NAMES = {
    "@cmcc": "中国移动",
    "@unicom": "中国联通",
    "@telecom": "中国电信",
    "@glgd": "中国广电",
}


class PortalPageState(enum.Enum):
    """Authentication state represented by a portal page."""

    ONLINE = "online"
    LOGIN_REQUIRED = "login_required"
    UNKNOWN = "unknown"


@dataclass(frozen=True)
class PortalLogoutRequest:
    """Logout request advertised by the portal page."""

    url: str
    method: str = "GET"
    data: str = ""


@dataclass(frozen=True)
class PortalSession:
    """Public session information exposed by the portal."""

    state: PortalPageState = PortalPageState.UNKNOWN
    user_id: str = ""
    ipv4: str = ""
    ipv6: str = ""
    logout_request: PortalLogoutRequest | None = None

    @property
    def is_online(self) -> bool:
        return self.state == PortalPageState.ONLINE

    @property
    def account_id(self) -> str:
        """Account portion before the operator suffix."""
        return split_portal_account(self.user_id)[0]

    @property
    def operator_name(self) -> str:
        """Chinese operator name derived from the public account suffix."""
        return split_portal_account(self.user_id)[1]


def split_portal_account(user_id: str) -> tuple[str, str]:
    """Split ``account@operator`` without exposing the suffix in account UI."""
    raw = user_id.strip()
    if not raw:
        return "", ""
    account, separator, suffix = raw.partition("@")
    if not separator:
        return account, "校园网"
    operator = PORTAL_OPERATOR_NAMES.get(
        f"@{suffix.lower()}",
        "未知运营商",
    )
    return account, operator


def parse_portal_page(html: str, portal_url: str) -> PortalSession:
    """Parse online state, public session details, and logout configuration."""
    title_match = re.search(r"<title[^>]*>(.*?)</title>", html, re.IGNORECASE | re.DOTALL)
    title = title_match.group(1).strip() if title_match else ""
    user_id = _script_value(html, "uid")
    ipv4 = _valid_ip_value(_script_value(html, "v4ip"))
    ipv6 = _valid_ip_value(_script_value(html, "v6ip"))

    # Only authoritative page-state markers prove that the campus session is
    # online. ``uid`` can remain populated briefly after logout and therefore
    # is session metadata, not a success condition by itself.
    online_markers = (
        "注销" in title,
        "Dr.COMWebLoginID_1.htm" in html,
        bool(re.search(r"\bpage\.run\s*\(\s*1\s*\)", html)),
    )
    login_markers = (
        "登录" in title and "成功" not in title,
        "Dr.COMWebLoginID_0.htm" in html,
        bool(re.search(r"\bpage\.run\s*\(\s*0\s*\)", html)),
        bool(re.search(r"<input[^>]+type\s*=\s*['\"]password['\"]", html, re.IGNORECASE)),
    )

    if any(online_markers):
        state = PortalPageState.ONLINE
    elif any(login_markers):
        state = PortalPageState.LOGIN_REQUIRED
    else:
        state = PortalPageState.UNKNOWN

    logout_request = _parse_logout_request(html, portal_url) if state == PortalPageState.ONLINE else None
    return PortalSession(
        state=state,
        user_id=user_id,
        ipv4=ipv4,
        ipv6=ipv6,
        logout_request=logout_request,
    )


def _parse_logout_request(html: str, portal_url: str) -> PortalLogoutRequest | None:
    path = _script_value(html, "authlogoutpath")
    if not path:
        return None

    if path.lower().startswith(("http://", "https://")):
        url = path
    else:
        portal = urlsplit(portal_url)
        host = _script_value(html, "authlogoutIP") or portal.hostname or ""
        port = _script_value(html, "authlogoutport")
        netloc = host
        if port and port not in ("0", "-1"):
            netloc = f"{host}:{port}"
        normalized_path = path if path.startswith("/") else f"/{path}"
        url = urlunsplit((portal.scheme or "http", netloc, normalized_path, "", ""))

    query = _script_value(html, "authlogoutparam").lstrip("?&")
    if query:
        separator = "&" if "?" in url else "?"
        url = f"{url}{separator}{query}"

    post_data = _script_value(html, "authlogoutpost")
    method = "POST" if post_data else "GET"
    return PortalLogoutRequest(url=url, method=method, data=post_data)


def _script_value(html: str, name: str) -> str:
    quoted = re.search(
        rf"\b{re.escape(name)}\s*=\s*(?P<quote>['\"])(?P<value>.*?)(?P=quote)\s*;",
        html,
        re.IGNORECASE | re.DOTALL,
    )
    if quoted:
        return quoted.group("value").strip()

    numeric = re.search(
        rf"\b{re.escape(name)}\s*=\s*(?P<value>-?\d+)\s*;",
        html,
        re.IGNORECASE,
    )
    return numeric.group("value").strip() if numeric else ""


def _valid_ip_value(value: str) -> str:
    value = value.strip()
    if not value or value in {"0.0.0.0", "000.000.000.000.", "::"}:
        return ""
    return value

File: src-tauri/src/portal.rs
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PortalPageState {
    Online,
    LoginRequired,
    #[default]
    Unknown,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PortalLogoutRequest {
    pub url: String,
    pub method: String,
    pub data: String,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PortalSession {
    pub state: PortalPageState,
    pub user_id: String,
    pub ipv4: String,
    pub ipv6: String,
    pub logout_request: Option<PortalLogoutRequest>,
}

impl PortalSession {
    pub fn is_online(&self) -> bool {
        self.state == PortalPageState::Online
    }

    pub fn account_id(&self) -> String {
        split_portal_account(&self.user_id).0
    }

    pub fn operator_name(&self) -> String {
        split_portal_account(&self.user_id).1
    }
}

pub fn split_portal_account(user_id: &str) -> (String, String) {
    let raw = user_id.trim();
    if raw.is_empty() {
        return (String::new(), String::new());
    }

    let Some((account, suffix)) = raw.split_once('@') else {
        return (raw.to_owned(), "校园网".to_owned());
    };
    let operator = match suffix.to_ascii_lowercase().as_str() {
        "cmcc" => "中国移动",
        "unicom" => "中国联通",
        "telecom" => "中国电信",
        "glgd" => "中国广电",
        _ => "未知运营商",
    };
    (account.to_owned(), operator.to_owned())
}

pub fn parse_portal_page(html: &str, portal_url: &str) -> PortalSession {
    let title = capture(html, r"(?is)<title[^>]*>(.*?)</title>").unwrap_or_default();
    let user_id = script_value(html, "uid");
    let ipv4 = valid_ip_value(&script_value(html, "v4ip"));
    let ipv6 = valid_ip_value(&script_value(html, "v6ip"));

    let online = title.contains("注销")
        || html.contains("Dr.COMWebLoginID_1.htm")
        || is_match(html, r"(?i)\bpage\.run\s*\(\s*1\s*\)");
    let login_required = (title.contains("登录") && !title.contains("成功"))
        || html.contains("Dr.COMWebLoginID_0.htm")
        || is_match(html, r"(?i)\bpage\.run\s*\(\s*0\s*\)")
        || is_match(html, r#"(?i)<input[^>]+type\s*=\s*['\"]password['\"]"#);

    let state = if online {
        PortalPageState::Online
    } else if login_required {
        PortalPageState::LoginRequired
    } else {
        PortalPageState::Unknown
    };
    let logout_request = (state == PortalPageState::Online)
        .then(|| parse_logout_request(html, portal_url))
        .flatten();

    PortalSession {
        state,
        user_id,
        ipv4,
        ipv6,
        logout_request,
    }
}

fn parse_logout_request(html: &str, portal_url: &str) -> Option<PortalLogoutRequest> {
    let path = script_value(html, "authlogoutpath");
    if path.is_empty() {
        return None;
    }

    let mut url = if path.starts_with("http://") || path.starts_with("https://") {
        path
    } else {
        let portal = url::Url::parse(portal_url).ok()?;
        let host = non_empty(&script_value(html, "authlogoutIP"))
            .unwrap_or_else(|| portal.host_str().unwrap_or_default().to_owned());
        let port = script_value(html, "authlogoutport");
        let authority = if port.is_empty() || matches!(port.as_str(), "0" | "-1") {
            host
        } else {
            format!("{host}:{port}")
        };
        let normalized_path = if path.starts_with('/') {
            path
        } else {
            format!("/{path}")
        };
        format!("{}://{authority}{normalized_path}", portal.scheme())
    };

    let query = script_value(html, "authlogoutparam")
        .trim_start_matches(['?', '&'])
        .to_owned();
    if !query.is_empty() {
        url.push(if url.contains('?') { '&' } else { '?' });
        url.push_str(&query);
    }

    let data = script_value(html, "authlogoutpost");
    let method = if data.is_empty() { "GET" } else { "POST" };
    Some(PortalLogoutRequest {
        url,
        method: method.to_owned(),
        data,
    })
}

fn script_value(html: &str, name: &str) -> String {
    let name = regex::escape(name);
    let quoted = format!(r#"(?is)\b{name}\s*=\s*['\"]([^'\"]*)['\"]\s*;"#);
    if let Some(value) = capture(html, &quoted) {
        return value;
    }
    let numeric = format!(r"(?i)\b{name}\s*=\s*(-?\d+)\s*;");
    capture(html, &numeric).unwrap_or_default()
}

fn valid_ip_value(value: &str) -> String {
    let value = value.trim();
    if value.is_empty() || matches!(value, "0.0.0.0" | "000.000.000.000." | "::") {
        String::new()
    } else {
        value.to_owned()
    }
}

fn non_empty(value: &str) -> Option<String> {
    (!value.is_empty()).then(|| value.to_owned())
}

fn capture(input: &str, pattern: &str) -> Option<String> {
    Regex::new(pattern)
        .ok()?
        .captures(input)?
        .get(1)
        .map(|value| value.as_str().trim().to_owned())
}

fn is_match(input: &str, pattern: &str) -> bool {
    Regex::new(pattern).is_ok_and(|regex| regex.is_match(input))
}

#[cfg(test)]
mod tests {
    use super::*;

    const ONLINE_PAGE: &str = r#"
        <title>注销页</title><!--Dr.COMWebLoginID_1.htm--><script>
        uid='000000000000@glgd';v4ip='192.0.2.10';v6ip='2001:db8::1';
        authlogoutIP='';authlogoutport=801;
        authlogoutpath='/eportal/?c=ACSetting&a=Logout&ver=1.0';
        authlogoutparam='url=drappall';authlogoutpost='';
        </script><script>page.run(1);</script>
    "#;

    #[test]
    fn parses_online_session_and_logout_request() {
        let session = parse_portal_page(ONLINE_PAGE, "http://10.0.1.5/");
        assert!(session.is_online());
        assert_eq!(session.account_id(), "000000000000");
        assert_eq!(session.operator_name(), "中国广电");
        assert_eq!(session.ipv4, "192.0.2.10");
        assert_eq!(session.ipv6, "2001:db8::1");
        let logout = session.logout_request.unwrap();
        assert_eq!(logout.method, "GET");
        assert_eq!(
            logout.url,
            "http://10.0.1.5:801/eportal/?c=ACSetting&a=Logout&ver=1.0&url=drappall"
        );
    }

    #[test]
    fn stale_user_id_does_not_override_login_state() {
        let html = "<title>登录页</title><input type='password'><script>uid='x@cmcc';page.run(0);</script>";
        let session = parse_portal_page(html, "http://10.0.1.5/");
        assert_eq!(session.state, PortalPageState::LoginRequired);
        assert!(!session.is_online());
    }

    #[test]
    fn splits_operator_suffixes() {
        assert_eq!(
            split_portal_account("20240001@cmcc"),
            ("20240001".into(), "中国移动".into())
        );
        assert_eq!(
            split_portal_account("20240001"),
            ("20240001".into(), "校园网".into())
        );
        assert_eq!(split_portal_account(""), (String::new(), String::new()));
    }
}
